src/main/java/com/amazon/opendistroforelasticsearch/knn/index/Parameter.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.knn.index;
+
+import org.elasticsearch.common.ValidationException;
+
+import java.util.function.Function;
+
+/**
+ * Parameter that can be set for a method component
+ *
+ * @param <T> Type parameter takes
+ */
+public abstract class Parameter<T> {
+
+    private T defaultValue;
+    private boolean inMethodString;
+    protected Function<T, Boolean> validator;
+
+    /**
+     * Constructor
+     *
+     * @param defaultValue of the parameter
+     * @param inMethodString whether the parameter is included in method string
+     */
+    public Parameter(T defaultValue, boolean inMethodString, Function<T, Boolean> validator) {
```

Comment:
Why not Predicate<T> instead of Function<T, Boolean>

Reply:
I did not know about Predicate. I will replace that.